File: leetcode-go/213_rob.go
package leetcode_go

func rob2(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	if len(nums) == 2 {
		return max(nums[0], nums[1])
	}
	result1 := robRange(nums, 0)
	result2 := robRange(nums, 1)
	return max(result1, result2)
}

func robRange(nums []int, start int) int {
	dp := make([]int, len(nums))
	dp[1] = nums[start]

	for i := 2; i < len(nums); i++ {
		dp[i] = max(dp[i-2]+nums[i-1+start], dp[i-1])
	}
	return dp[len(nums)-1]
}

File: src/dp/_213_rob.rs
struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        fn rob_range(nums: &[i32], start: usize) -> i32 {
            let mut dp = vec![0; nums.len()];
            dp[1] = nums[start];
            for i in 2..nums.len() {
                dp[i] = dp[i - 1].max(dp[i - 2] + nums[i - 1 + start])
            }
            return dp[nums.len()];
        }
        if nums.len() == 1 {
            return nums[0];
        }
        if nums.len() == 2 {
            return nums[0].max(nums[1]);
        }
        rob_range(&nums, 0).max(rob_range(&nums, 1))
    }
}